### DISCUSSION TITLE
Non-blocking workflow execution

### DISCUSSION THREAD

**Discussion by iparask** [radical-cybertools/radical.entk#542] [Ideas]
From @euhruska:

Feature request for entk to enable adding individual non-blocking tasks.

Two proposed new function calls (to be discussed):

appman.run_non_blocking()
Description: Starts an Application manager with an pipeline. Pipeline might contain only an empty stage. Keep running even if no tasks are left to be run until walltime runs out. This call has to be non-blocking.

appman.run_task_nonblocking(task)
Description: Add to above Application manager one task. This call has to be non-blocking. Will be executed as soon as possible. No data transfer from tasks back except task.state.

Mock code:
https://github.com/radical-cybertools/radical.midas/blob/master/midas_mock_non_blocking.py

Schema:
https://docs.google.com/presentation/d/1uXyYE10QkU33r6ucnk_PSScGLrsWbLccd4WSiG61bs8

**Comment by iparask** [comment]
My notes:

1. Currently, EnTK supports the execution of tasks where any logic that creates a dependency between tasks, apart from temporal, is defined with the task.
2. The definition of such logic is done at creating an application, whether it is just an EnTK script or a python executable. As a result, the logic is considered static.
3. Such logic can be executed while a stage is executing.
4. As a result, the execution of a workflow should be non-blocking.
5. As a workflow is executing, specifically a stage, those may change based on the logic's results.
6. Provide access to python's interpreter terminal, and thus point 4, to the user to do analysis on the client-side and push new tasks on pipelines and stages.
7. We need to be able to probe the state of a task during runtime.
8. Do we need to move data? No, since this use case is for systems like Summit where the execution is launched from the login node, and as such, the python interpreter already has access to the shared filesystem.
9. We need to pull information from a task as it is executing. What type of information?
10. Stressed that our systems are libraries and, as such, the user is responsible for any interactive use.

**Comment by mturilli** [comment]
> appman.run_non_blocking()
> Description: Starts an Application manager with an pipeline. Pipeline might contain only an empty stage. Keep running even if no tasks are left to be run until walltime runs out. This call has to be non-blocking.

`appman.run_non_blocking()` would be equivalent to submit a pilot with no tasks, i.e., submitting a job to an HPC machine running only RP Agent without running any task. As we pay (allocation and, ultimately, money) whether we run a task or not, we should wait to submit a pilot until we have at least one task to execute. We could still specify the Pipeline and the stage, do whatever we need to do before being able to specify the first task, and then add that first task to the stage. At that point, we could submit the pipeline for execution, triggering the submission of a pilot.
 
> appman.run_task_nonblocking(task)
> Description: Add to above Application manager one task. This call has to be non-blocking. Will be executed as soon as possible. No data transfer from tasks back except task.state.

If we follow the above, I would propose to have something like:

```
p = Pipeline()
s = Stage()
t1 = Task()

t1.name = 'my-first-task'
t1.executable = '/bin/echo'
t1.arguments = ['Hello World 1']

s.add_tasks(t1)
p.add_stages(s)

appman = AppManager(hostname=hostname, port=port, username=username, password=password)
res_dict = {'resource': 'local.localhost', 'walltime': 10, 'cpus': 1}
appman.resource_desc = res_dict
appman.workflow = p
appman.run_nonblocking(terminate=False)

t2.name = 'my-first-task'
t2.executable = '/bin/echo'
t2.arguments = ['Hello World 2']

s.add_tasks(t2)
```

That would imply that, if `appman` has submitted `t1`, `appman` would submit `t2` as soon as it is added, avoiding the need of a dedicated new call to run a task. `Terminate=false` would enable `appman` to keep the pilot alive even if `t1` has finished to execute before `t2` is submitted. 

As discussed, there are some implications on EnTK's state model. Currently, `p` would be marked as `Done` if `t2` is submitted after `t1` has ended. We can discuss different approaches to this, also taking into account what we learned about pausing/restarting/terminating pipelines in REPEX. For example, should we add an explicit call to terminate a pipeline and stages? We could replace `appman.workflow` with:

```
appman.add_workflow = p
...
appman.del_workflow = p
```

In that way, `p` would not be marked as terminated until `del_workflow` is called. We would have to automatically move `p` to a final state  in some conditions, e.g., explicit pilot termination. Note that we would not do that when a pilot crashes and `appman` has been configured to resubmit that pilot. We could do the same with stages and manage automatic termination with priorities between p and s, i.e., terminate a state if its pipeline has reached a final state.

There are many details/conditions I am not explicitly addressing here and there are other approaches to consider but, if you like the approach I suggested, I will be happy to work together towards a full spec.

**Reply by iparask** [reply]
> > appman.run_non_blocking()
> > Description: Starts an Application manager with an pipeline. Pipeline might contain only an empty stage. Keep running even if no tasks are left to be run until walltime runs out. This call has to be non-blocking.
> 
> `appman.run_non_blocking()` would be equivalent to submit a pilot with no tasks, i.e., submitting a job to an HPC machine running only RP Agent without running any task. As we pay (allocation and, ultimately, money) whether we run a task or not, we should wait to submit a pilot until we have at least one task to execute. We could still specify the Pipeline and the stage, do whatever we need to do before being able to specify the first task, and then add that first task to the stage. At that point, we could submit the pipeline for execution, triggering the submission of a pilot.

I agree here with @mturilli. Also, @euhruska, do you assume that you will have access to the requested resources immediately? 

> 
> > appman.run_task_nonblocking(task)
> > Description: Add to above Application manager one task. This call has to be non-blocking. Will be executed as soon as possible. No data transfer from tasks back except task.state.
> 
> If we follow the above, I would propose to have something like:
> 
> ```
> p = Pipeline()
> s = Stage()
> t1 = Task()
> 
> t1.name = 'my-first-task'
> t1.executable = '/bin/echo'
> t1.arguments = ['Hello World 1']
> 
> s.add_tasks(t1)
> p.add_stages(s)
> 
> appman = AppManager(hostname=hostname, port=port, username=username, password=password)
> res_dict = {'resource': 'local.localhost', 'walltime': 10, 'cpus': 1}
> appman.resource_desc = res_dict
> appman.workflow = p
> appman.run_nonblocking(terminate=False)
> 
> t2.name = 'my-first-task'
> t2.executable = '/bin/echo'
> t2.arguments = ['Hello World 2']
> 
> s.add_tasks(t2)
> ```
> 
> That would imply that, if `appman` has submitted `t1`, `appman` would submit `t2` as soon as it is added, avoiding the need of a dedicated new call to run a task. `Terminate=false` would enable `appman` to keep the pilot alive even if `t1` has finished to execute before `t2` is submitted.
> 
> As discussed, there are some implications on EnTK's state model. Currently, `p` would be marked as `Done` if `t2` is submitted after `t1` has ended. We can discuss different approaches to this, also taking into account what we learned about pausing/restarting/terminating pipelines in REPEX. For example, should we add an explicit call to terminate a pipeline and stages?

@mturilli what do you mean here exactly? I think if a pipeline is marked as done, then a new pipeline will be submitted with the new task. In addition, I would like to suggest that if a stage, or pipeline is marked `Done` no more tasks can be added to that stage, or pipeline.

We could replace `appman.workflow` with:
> 
> ```
> appman.add_workflow = p
> ...

I like `add_pipeline` for adding pipelines to a workflow.

> appman.del_workflow = p
> ```
> 
> In that way, `p` would not be marked as terminated until `del_workflow` is called. We would have to automatically move `p` to a final state in some conditions, e.g., explicit pilot termination. Note that we would not do that when a pilot crashes and `appman` has been configured to resubmit that pilot. We could do the same with stages and manage automatic termination with priorities between p and s, i.e., terminate a state if its pipeline has reached a final state.

I would suggest `p.terminate` which keeps the pipeline in the workflow, and moves it to the final state. I am not sure it helps with stages as stages define the temporal dependency between tasks. In my mind, stages are executed and if finished they move to a `Done` state. The pipeline though can remain open and produce an error if the task is added to a stage that is in the `Done` state.

**Reply by andre-merzky** [reply]
> appman.add_workflow = p

This is a cardinal sin in RE's API: all those side effects on _*property setters*_ really make, IMO, the API difficult to use and to understand.  With the above, you set a _*property*_.  What happens to that `p`?  When is it added / used?  When is an _*action*_ triggered on it?  What happens if I set it twice?  How can I 'unset' it?  

If you want to perform the action of submitting a workflow, then the API should express the action 'submit a workflow', i.e., 
```py
appman.submit_workflow(p)
```

Next: why is the _*app*_ manager not managing _*app*_s?  It is a workflow manager - shouldn't it be named like it?

Also: if you want to have the API non-blocking, and at the same time want it to handle multiple workflows, then a workflow must be a first order citizen in the API.  If we touch the RE API, then I would suggest to really change it, instead of trying to shoehorn async semantics on the side effects of the current property setters... *shiver*

I worked over my [API proposal](https://github.com/radical-cybertools/radical.entk/tree/feature/api/src/radical/entk/api.v2) again to make it a bit more consistent with the points above...

**Reply by euhruska** [reply]
I don't think multiple workflows is necessary for this feature request. The API refactoring sounds interesting but might be separate. My usecase would only use one single stage and one single pipeline and one single workflow. My view is that pipelines and stages would not be marked Done or terminated, and only closed when appman is terminated. Execution barriers could be set by user by waiting to submit new tasks. At least this might be first step of implementation, the second step might be supporting multiple stages and workflows.
Maybe instead of appman.del_workflow = p and p.terminate have appman.terminate(), which closes everything?

Terminate=false sounds good, I would even suggest appman.run(blocking=False, terminate=False).

@iparask The requested resources might not be immediately available. If it simplifies the implementation we could assume resources are available immediately, that would be sufficient. If the resources are not immediately available then starting the job without tasks might make sense due to queue wait time.

**Reply by andre-merzky** [reply]
> I don't think multiple workflows is necessary for this feature request.

Agree - that was more for the overall API discussion.

> My view is that pipelines and stages would not be marked Done or terminated

This is difficult for dependency resolution: if stage 2 runs after stage 1, what does it mean if a task is added to stage 1 after stage 2 started?  Should it restart?  Should data be staged again (as they may have changed)? When can stage 2 start in the first place if stage 1 is never `DONE`?  How would a user know that a stage is `DONE` - would he track tasks in stages and pipelines themselves?  Why would they use ENTK then in the first place?

So this is unlikely to happen, at least not in the simple sense, as the consequences for the overall workflow semantics are too confusing.  If tasks is the only abstraction you need, then you may want to consider using radical.pilot directly.

> @iparask The requested resources might not be immediately available. If it simplifies the implementation we could assume resources are available immediately, that would be sufficient. If the resources are not immediately available then starting the job without tasks might make sense due to queue wait time.

I agree with that: I don't see any benefit for waiting with any action while the pilot job is in the queue.  Specifically:

> appman.run_non_blocking() would be equivalent to submit a pilot with no tasks, i.e., submitting a job to an HPC machine running only RP Agent without running any task. As we pay (allocation and, ultimately, money) whether we run a task or not, we should wait to submit a pilot until we have at least one task to execute.

I don't think this is a real concern: When the application requests resources, we should begin the process of providing them.  It is the responsibility of the user to correctly size and time the resources anyway, and the starting point is no different IMHO.  Adding stages and tasks to the workflow will be orders of magnitudes faster than pilot provisioning anyway, and I don't see a realistic use case where the application starts the workflow execution and then waits for hours until it submits the first task.

**Reply by iparask** [reply]
> If you want to perform the action of submitting a workflow, then the API should express the action 'submit a workflow', i.e.,
> 
> ```python
> appman.submit_workflow(p)
> ```
>
> Next: why is the __app__ manager not managing __app__s? It is a workflow manager - shouldn't it be named like it?
> 

I think it might be better to add pipelines to a workflow. Somoething like:
```python
appman.add_pipeline(p)
```

I am not sure how to answer the first question and I think you have a valid point there.

> Also: if you want to have the API non-blocking, and at the same time want it to handle multiple workflows, then a workflow must be a first order citizen in the API. If we touch the RE API, then I would suggest to really change it, instead of trying to shoehorn async semantics on the side effects of the current property setters... _shiver_

Let me read a bit more about properties and I'll get back at you.

**Reply by iparask** [reply]
> > I don't think multiple workflows is necessary for this feature request.
> 
> Agree - that was more for the overall API discussion.

Multiple workflows may be confusing, but I think supporting a workflow that has multiple pipelines with multiple stages and tasks is necessary. 

> 
> > My view is that pipelines and stages would not be marked Done or terminated
> 
> This is difficult for dependency resolution: if stage 2 runs after stage 1, what does it mean if a task is added to stage 1 after stage 2 started? 
>Should it restart? Should data be staged again (as they may have changed)? When can stage 2 start in the first place if stage 1 is never `DONE`?

My proposal is that is a stage has no more tasks that are waiting to execute to be marked as `Done`. If the user tries to add a new task in that stage, EnTK issues an error and requests the user to add a stage. I do not see a reason to restart, as we destroy temporal dependency introduced by stages. 

> How would a user know that a stage is `DONE` - would he track tasks in stages and pipelines themselves? 

We can keep pipelines from transitioning to the `Done` state when async execution is enabled. I suggest the following: 
1. tasks and stages keep the current state model they have.
2. Pipelines transition to the done state if all stages are in `DONE` state and the execution is non-blocking. If the execution is non-blocking, pipelines do not transition to a final state, unless users call something like `p.close` or close the whole execution.

**Reply by euhruska** [reply]
>  Why would they use ENTK then in the first place?

What would be the disadvantages of using radical pilot, besides not having stages and having to track state of pipelines yourself?

**Reply by andre-merzky** [reply]
> > Why would they use ENTK then in the first place?
> 
> What would be the disadvantages of using radical pilot, besides not having stages and having to track state of pipelines yourself?

Hi @euhruska: I don't know your specific use case, and others may me more qualified to respond.  In general terms: 

The main purpose of EnTK is to add structure to your workload, to simplify data management and task orchestration.  RP OTOH is bare-bones: no task dependencies, no tracking of data dependencies, no notion of collections of task.  RP really boils down to : submit a pilot, run `n` tasks on it, done.  

Having said that: RP runs below EnTK as you know, and as such provides all semantics used by EnTK tasks also: similar task descriptions, data staging, state event callbacks, etc.  It is in general a bit faster, but not so much that it makes a difference in most cases.

**Reply by euhruska** [reply]
I think RP is sufficient for this use case and the required task dependencies would be easier to implement in RP than in EnTK. Do we close this discussion?

**Reply by mturilli** [reply]
I am afraid adding non-blocking calls in RP is not in the current roadmap so I would stick to EnTK. @andre-merzky  would you confirm that? @euhruska in our last call, my notes tell me that we explicitly discuss the necessity of representing dependencies among tasks. I am not sure I wrote it down correctly, but I noted that you had 4 types of tasks that needed to execute one after the other. Did I get it wrong?

**Reply by euhruska** [reply]
umgr.submit_units() appears to behave non-blocking and can be called for each task separately, just ran a mock code demonstrating this functionality. @mturilli The dependencies would be executed outside of EnTK/RP, between the non-blocking calls. There would be just one type of tasks from EnTK/RP perspective, but the user could define each task as desired.

**Reply by mturilli** [reply]
If RP and `umgr.submit_units()` work for you happy to go that way. The whole idea of RCT is to enable to users not to have to code the abstractions that each RCT provides. For example, in this case you seem to wanting to write code to manage task dependencies. The idea of EnTK is to provide those capabilities for you. This is why we would be interested anyway to see how we would have to use EnTK to satisfy your use case.

**Reply by euhruska** [reply]
Agreed. The requirement of non-blocking task submission made EnTK not currently applicable. Once the use case is demonstrated on RP level or if the task dependencies would be too complex, we still can follow the above discussed approach to support non-blocking tasks in EnTK.

**Comment by iparask** [comment]
A question just came to my mind. Will we allow to mix blocking and non-blocking execution? Something like:

```python
appman.submit_pipelines(set([p]))
appman.run(blocking=False)
... I'm doing something

appman.submit_pipelines(set([p]))
appman.run(blocking=True)
```
